### PR TITLE
Feature/simple socket

### PR DESCRIPTION
**Pull request by alfonso-soto**

### CODE REVIEW

**Comment by alfonso-soto** [comment]
Change the name of the class -> RTIRawTransportImpl

File: srcCpp/RTISocketImpl.cxx [hunk 1]
```diff
@@ -0,0 +1,1372 @@
...
+/*********************************************************
+ * PrintCmdLineHelp
+ */
+void RTISocketImpl::PrintCmdLineHelp() {
+    /**************************************************************************/
+    std::string usage_string = std::string(
+            "\t-sendQueueSize <number>       - Sets number of samples (or batches) in send\n") +
+            "\t                                queue, default 50\n" +
+            "\t-domain <ID>                  - RTI DDS Domain, default 1\n" +
+            "\t-bestEffort                   - Run test in best effort mode, default reliable\n"
```
Comment (javmorcas) [inline srcCpp/RTISocketImpl.cxx]: Not needed.

File: srcCpp/RTISocketImpl.cxx [hunk 2]
```diff
@@ -0,0 +1,1372 @@
...
+            "\t                                " +
+            "\t                                Default: \"default\" (If using asynchronous).\n" +
+            "\t-peer <address>               - Adds a peer to the peer host address list.\n" +
+            "\t                                This argument may be repeated to indicate multiple peers\n" +
+            "\t-noBlockingSockets            - Control blocking behavior of send sockets to never block.\n" +
```
Comment (javmorcas) [inline srcCpp/RTISocketImpl.cxx]: Create this for RTIDDSImpl as well (new feature)

File: srcCpp/RTISocketImpl.cxx [hunk 3]
```diff
@@ -0,0 +1,1372 @@
...
+            "\t                                CHANGING THIS FROM THE DEFAULT CAN CAUSE SIGNIFICANT PERFORMANCE PROBLEMS.\n" +
+            "\t-socketsBasePort              - Changes the base port from where the 3 ports\n"
```
Comment (javmorcas) [inline srcCpp/RTISocketImpl.cxx]: Remove this feature.

**Comment by javmorcas** [comment]
**1 to Many** Discussion:

- We want to support 1 to many



Ports:

Pub:
ParticipantID = PubID

Sub:
ParticipantID = NumPublishersID + 2*(SubscriberID)

File: srcCpp/perftest_publisher.cxx [hunk 4]
```diff
@@ -235,6 +242,8 @@ bool perftest_cpp::ParseConfig(int argc, char *argv[])
 {
     _MessagingArgc = 0;
     _MessagingArgv = new char*[argc];
+    int _MessagingSocketsArgc = 0;
+    char **_MessagingSocketsArgv = new char*[argc];
 
     if (_MessagingArgv == NULL) {
```
Comment (javmorcas) [inline srcCpp/perftest_publisher.cxx]: Remove if (_MessagingArgv == NULL) {

Initialize the _MessagingSocketsArgv

File: srcCpp/perftest_publisher.cxx [hunk 5]
```diff
@@ -518,6 +542,11 @@ bool perftest_cpp::ParseConfig(int argc, char *argv[])
                 return false;
             }
             _NumSubscribers = strtol(argv[i], NULL, 10);
+            _MessagingSocketsArgv[_MessagingSocketsArgc] =
+                    DDS_String_dup(argv[i - 1]);
+            _MessagingSocketsArgv[++_MessagingSocketsArgc] =
```
Comment (javmorcas) [inline srcCpp/perftest_publisher.cxx]: You can use _MessagingSocketsArgc++ in the previous string and just do _MessagingSocketsArgc and not ++_MessagingSocketsArgc

File: srcCpp/perftest_publisher.cxx [hunk 6]
```diff
@@ -834,6 +868,26 @@ bool perftest_cpp::ParseConfig(int argc, char *argv[])
         }
     }
 
+    if (_useSockets) {
+        int count = 0;
```
Comment (javmorcas) [inline srcCpp/perftest_publisher.cxx]: What is the reason?

File: srcCpp/RTISocketImpl.cxx [hunk 7]
```diff
@@ -0,0 +1,1282 @@
...
+    }
+
+    if (_transport.useMulticast && (_peer_host_count > 0)) {
+        fprintf(stderr,
+                "\tFor multicast, set only one Subscriber and one Publisher\n");
+        return false;
+    }
+
+    return true;
+}
```
Comment (alfonso-soto) [inline srcCpp/RTISocketImpl.cxx]: Create a new parseConf()

File: srcCpp/RTISocketImpl.cxx [hunk 8]
```diff
@@ -0,0 +1,1282 @@
...
+
+    // Set default interface
+    _transport.allowInterfaces = std::string("127.0.0.1");
+
+}
+
+/*********************************************************
+ * Shutdown
+ */
+void RTISocketImpl::Shutdown()
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTISocketImpl.cxx]: All of this functions are void, so no error check possible.

File: srcCpp/RTISocketImpl.cxx [hunk 9]
```diff
@@ -0,0 +1,1282 @@
...
+    usage_string += _transport.helpMessageString();
+
+    fprintf(stderr, "%s", usage_string.c_str());
+}
+
+/*********************************************************
+ * ParseConfig
+ */
+bool RTISocketImpl::ParseConfig(int argc, char *argv[]) {
+    unsigned long _scan_max_size = 0;
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTISocketImpl.cxx]: Fixed

File: srcCpp/RTISocketImpl.cxx [hunk 10]
```diff
@@ -0,0 +1,1282 @@
...
+    no_socket_params_v.push_back(std::string("-transportWanId"));
+    no_socket_params_v.push_back(std::string("-transportSecureWan"));
+
+    std::string params_info = std::string(
+            "Parameters not supported by sockets: (Delete them and try again)\n");
+
+    /* Print all the non compatibles params together and return */
+    bool found = false;
+    std::vector<std::string>::iterator it;
+    for (int i = 0; i < argc; ++i ) {
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTISocketImpl.cxx]: The two 'i' are for different purpose, so I decided to change the name to the variable.

File: srcCpp/RTISocketImpl.cxx [hunk 11]
```diff
@@ -0,0 +1,1282 @@
...
+    no_socket_params_v.push_back(std::string("-transportWanServerAddress"));
+    no_socket_params_v.push_back(std::string("-transportWanServerPort"));
+    no_socket_params_v.push_back(std::string("-transportWanId"));
+    no_socket_params_v.push_back(std::string("-transportSecureWan"));
+
+    std::string params_info = std::string(
+            "Parameters not supported by sockets: (Delete them and try again)\n");
+
+    /* Print all the non compatibles params together and return */
+    bool found = false;
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTISocketImpl.cxx]: Done

File: srcCpp/RTISocketImpl.cxx (hunk 7 above)
Comment (alfonso-soto) [reply srcCpp/RTISocketImpl.cxx]: Postponed, Wait until the parse class it's done.

File: srcCpp/RTIRawTransportImpl.cxx [hunk 12]
```diff
@@ -0,0 +1,1481 @@
...
+    for (int i = 0; i < plugin->_interfacesCount; i++) {
+        if (plugin->_interfaceArray[i]._interfaceFlags
+            & RTI_OSAPI_SOCKET_INTERFACE_FLAG_MULTICAST) {
+            ++count;
+        }
+    }
+    return count;
+}
+
+struct REDAWorker *RawTransport_get_worker_per_threadI(
```
Comment (alfonso-soto) [inline srcCpp/RTIRawTransportImpl.cxx]: Change name, RawTrasnportGetWorkerPerThread

Discuss with Jose and Javi Worker creations.

File: srcCpp/RTIRawTransportImpl.cxx [hunk 13]
```diff
@@ -0,0 +1,1481 @@
...
+
+    void Flush()
+    {
+        /* If there is no data, no need to flush */
+        if (_sendBuffer.length == 0) {
+            return;
+        }
+
+        if (!SendMessage()) {
+            /* No need of print error. This will be represented as loss packets */
```
Comment (alfonso-soto) [inline srcCpp/RTIRawTransportImpl.cxx]: lost

Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: Remove the If statement. Just set the comment

File: srcCpp/RTIRawTransportImpl.cxx [hunk 14]
```diff
@@ -0,0 +1,1481 @@
...
+        }
+
+        /* If the send it's done, reset the stream to fill the buffer again */
```
Comment (alfonso-soto) [inline srcCpp/RTIRawTransportImpl.cxx]: If the send is....

File: srcCpp/RTIRawTransportImpl.cxx [hunk 15]
```diff
@@ -0,0 +1,1481 @@
...
+        }
+    }
+
+    TestMessage *ReceiveMessage() {
+
+
+        while (true) {
+            if (_noData) {
+                if (_worker == NULL){
+                    _worker = RawTransport_get_worker_per_threadI(
```
Comment (alfonso-soto) [inline srcCpp/RTIRawTransportImpl.cxx]: Put the _worker = ... outside of the while

File: srcCpp/RTIRawTransportImpl.cxx [hunk 16]
```diff
@@ -0,0 +1,1481 @@
...
+
+RTIOsapiThreadTssFactory *RTIRawTransportImpl::getTssFactory()
+{
+    return _tssFactory;
+}
+
+/*********************************************************
+ * SupportFunctions
+ */
+bool RTIRawTransportImpl::SupportListener()
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: Done

File: srcCpp/RTIRawTransportImpl.cxx [hunk 17]
```diff
@@ -0,0 +1,1481 @@
...
+    }
+
+    void Shutdown()
+    {
+        if (_sendBuffer.pointer != NULL) {
+            RTIOsapiHeap_freeBuffer(_sendBuffer.pointer);
+            _sendBuffer.pointer = NULL;
+        }
+
+        if (_parent->getWorkerFactory() != NULL && _worker != NULL) {
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: Done.

File: srcCpp/RTIRawTransportImpl.cxx (hunk 12 above)
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: Name fixed.

File: srcCpp/RTIRawTransportImpl.cxx (hunk 13 above)
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: Done.

File: srcCpp/RTIRawTransportImpl.cxx (hunk 14 above)
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: Done.

File: srcCpp/RTIRawTransportImpl.cxx (hunk 15 above)
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: Done

File: srcCpp/RTIDDSImpl.h [hunk 18]
```diff
@@ -113,31 +66,49 @@ class RTIDDSImpl : public IMessaging
...
     IMessagingReader *CreateReader(const char *topic_name, IMessagingCB *callback);
 
     DDSTopicDescription *CreateCft(const char *topic_name, DDSTopic *topic);
 
+    /*
+     * Those two functions calculate the time cost on serialization,
+     * with a precision of microseconds.
+     */
+
+    static double ObtainDDSSerializeTimeCost(
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIDDSImpl.h]: :ballot_box_with_check:

File: srcCpp/RTIDDSImpl.cxx [hunk 19]
```diff
@@ -2651,21 +2196,203 @@ unsigned long RTIDDSImpl<T>::GetInitializationSampleCount()
...
+        unsigned int iters)
+{
+    T data;
+    double timeInit = 0;
+    double timeFinish = 0;
+    double serializeTime;
+
+    unsigned int maxSizeSerializedSample = 0;
+    char *buffer;
+    char *serializeBuffer;
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIDDSImpl.cxx]: :ballot_box_with_check:

File: srcCpp/RTIDDSImpl.cxx [hunk 20]
```diff
@@ -2651,21 +2196,203 @@ unsigned long RTIDDSImpl<T>::GetInitializationSampleCount()
...
+    timeInit = (unsigned int) perftest_cpp::GetTimeUsec();
+
+    for (unsigned int i = 0; i < iters; i++) {
+        if (DDS_RETCODE_OK != T::TypeSupport::serialize_data_to_cdr_buffer(
+                serializeBuffer,
+                maxSizeSerializedSample,
+                &data)){
+            fprintf(stderr,
+                    "Fail to serialize sample on ObtainDDSSerializeTimeCost\n");
+            return 0;
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIDDSImpl.cxx]: :ballot_box_with_check:

File: srcCpp/perftest_publisher.cxx [hunk 21]
```diff
@@ -882,88 +406,99 @@ void perftest_cpp::PrintConfiguration()
...
+            stringStream << "\tNumber of samples: "
+                         << _PM.get<unsigned long long>("numIter")
+                         << "\n";
         }
     }
 
     // Listener/WaitSets
     stringStream << "\tReceive using: ";
-    if (_UseReadThread) {
+    if (_PM.get<bool>("useReadThread") && !_PM.get<bool>("rawTransport")) {
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/perftest_publisher.cxx]: :ballot_box_with_check:

File: srcCpp/perftest_publisher.cxx [hunk 22]
```diff
@@ -1595,6 +1161,40 @@ class LatencyListener : public IMessagingCB
...
+                        ObtainDDSSerializeTimeCost(totalSampleSize);
+                deserializeTime = RTIDDSImpl<TestDataKeyedLarge_t>::
+                        ObtainDDSDeserializeTimeCost(totalSampleSize);
+            } else {
+                serializeTime = RTIDDSImpl<TestDataLarge_t>::
+                        ObtainDDSSerializeTimeCost(totalSampleSize);
+                deserializeTime = RTIDDSImpl<TestDataLarge_t>::
+                        ObtainDDSDeserializeTimeCost(totalSampleSize);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/perftest_publisher.cxx]: :ballot_box_with_check:

File: srcCpp/perftest_publisher.cxx [hunk 23]
```diff
@@ -2164,28 +1845,39 @@ int perftest_cpp::Publisher()
...
 
     if (writer != NULL) {
-        delete(writer);
+        delete (writer);
     }
 
-    if (announcement_reader != NULL) {
-        delete(announcement_reader);
+    if (reader_listener != NULL) {
+        delete (reader_listener);
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/perftest_publisher.cxx]: :ballot_box_with_check:

File: srcCpp/RTIRawTransportImpl.h [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+#include "perftestSupport.h"
+#include "perftest_cpp.h"
+#include "transport/transport_socketutil.h"
+#include "transport/transport_udpv4.h"
+
+#include "perftestPlugin.h"
+
+#define RTIPERFTEST_MAX_PEERS 1024
+
+class PeerData;
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.h]: :ballot_box_with_check:

File: srcCpp/RTIRawTransportImpl.h [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+  public:
+    RTIRawTransportImpl();
+    ~RTIRawTransportImpl() { Shutdown(); }
+    bool validate_input();
+    std::string PrintConfiguration();
+    bool Initialize(ParameterManager &PM);
+    void Shutdown();
+    int GetBatchSize();
+
+    bool SupportsListener();
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.h]: :ballot_box_with_check:

File: srcCpp/RTIRawTransportImpl.cxx [hunk 26]
```diff
@@ -0,0 +1,1311 @@
...
+
+RTIOsapiThreadTssFactory *RTIRawTransportImpl::get_tss_factory()
+{
+    return _tssFactory;
+}
+
+ParameterManager *RTIRawTransportImpl::get_parameter_manager()
+{
+    return _PM;
+}
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: :ballot_box_with_check:

File: srcCpp/RTIRawTransportImpl.h [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+    bool SupportsDiscovery();
+
+    IMessagingWriter *CreateWriter(const char *topic_name);
+
+    /* This implementation dont support listener so callback is ignore. */
+    IMessagingReader *
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.h]: :ballot_box_with_check:
The start should be with the returned value. So it's ok like this.

File: srcCpp/RTIRawTransportImpl.h [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+    struct REDAWorkerFactory *get_worker_factory();
+    RTIOsapiThreadTssFactory *get_tss_factory();
+    ParameterManager *get_parameter_manager();
+
+    /* Calculate the port depending of the Id of the subscriber.*/
+    unsigned int
+    get_send_unicast_port(const char *topicName, unsigned int subId = 0);
+
+    /* Get the multicast address that match to the topic name */
+    bool getMulticastTransportAddr(
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.h]: :ballot_box_with_check:

File: srcCpp/RTIRawTransportImpl.cxx [hunk 29]
```diff
@@ -0,0 +1,1311 @@
...
+          _pongSemaphore(NULL),
+          _plugin(NULL),
+          _workerFactory(NULL),
+          _exclusiveArea(NULL),
+          _tssFactory(NULL),
+          _PM(NULL)
+{
+
+    // Reserve space for the peers
+    PeerData::resourcesList.reserve(RTIPERFTEST_MAX_PEERS);
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: Removed
:ballot_box_with_check:

File: srcCpp/RTIRawTransportImpl.cxx [hunk 30]
```diff
@@ -0,0 +1,1311 @@
...
+        }
+    }
+
+    // Manage transport parameter
+    if (!_transport.validate_input()) {
+        fprintf(stderr, "Failure validation the transport options.\n");
+        return false;
+    };
+
+    // Manage parameter -peerRT
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: :ballot_box_with_check:

[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: Update documentation regarding -peer Paramter

File: srcCpp/perftest_publisher.cxx [hunk 31]
```diff
@@ -1595,6 +1166,40 @@ class LatencyListener : public IMessagingCB
         );
         fflush(stdout);
 
+        if (!unbounded) {
+            if (isKeyed) {
+                serializeTime = RTIDDSImpl<TestDataKeyed_t>::
+                        ObtainDDSSerializeTimeCost(totalSampleSize);
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/perftest_publisher.cxx]: :ballot_box_with_check:

File: srcCpp/RTIRawTransportImpl.cxx [hunk 32]
```diff
@@ -0,0 +1,1257 @@
...
+            if (_readThreadSemaphore != NULL) {
+                if (RTIOsapiSemaphore_give(_readThreadSemaphore)
+                        != RTI_OSAPI_SEMAPHORE_STATUS_OK) {
+                    fprintf(stderr, "Unexpected error giving semaphore\n");
+                    return;
+                }
+            }
+        }
+
+        if (_readThreadSemaphore != NULL) {
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: :ballot_box_with_check:

File: srcCpp/RTIRawTransportImpl.cxx [hunk 33]
```diff
@@ -0,0 +1,1257 @@
...
+            _message.size = _data.bin_data.length();
+            _message.data = (char *)_data.bin_data.get_contiguous_bufferI();
+
+            return &_message;
+
+        }
+
+    }
+
+    struct RTIOsapiSemaphore *GetReadThreadSemaphore()
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIRawTransportImpl.cxx]: :ballot_box_with_check:

File: srcCpp/perftest_publisher.cxx [hunk 34]
```diff
@@ -1744,18 +1348,28 @@ class LatencyListener : public IMessagingCB
...
+static void *ReadThread(void *arg)
 {
-    LatencyListener *listener = static_cast<LatencyListener *>(arg);
+    ListenerType *listener = static_cast<ListenerType *>(arg);
     TestMessage *message = NULL;
+    RTIOsapiSemaphore *semaphore = listener->_reader->GetReadThreadSemaphore();
+
+    // Meanwhile it's been use the reader, avoid the destruction of it.
+    if (semaphore != NULL) {
+        if (RTIOsapiSemaphore_take(semaphore, NULL)
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/perftest_publisher.cxx]: :ballot_box_with_check:

File: srcCpp/MessagingIF.h [hunk 35]
```diff
@@ -41,10 +43,42 @@ class IMessagingCB
...
+        }
+    }
+
     virtual void ProcessMessage(TestMessage &message) = 0;
+
+    /*
+     * Default constructor, if a inheritance class declare this function must
+     * also call to delete_sync_semaphore().
+     */
+    virtual ~IMessagingCB() {
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/MessagingIF.h]: If the destructor of inherited class is called, the destructot of the base class will be also called.
:ballot_box_with_check:

File: srcCpp/perftest_publisher.cxx [hunk 36]
```diff
@@ -1210,20 +748,46 @@ class ThroughputListener : public IMessagingCB
...
+        * To support -latencyTest plus -useReadThread we need to signal
+        * --HERE-- the internal semaphore used in RTIDDSImpl.cxx as
+        * we now do in the listener on_data_available callback
+        * inside RTIDDSImpl.cxx
+        *
+        */
+    }
+
+    // This will allow the main thread to delete the thread that used this function.
+    if (semaphore != NULL) {
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/perftest_publisher.cxx]: :ballot_box_with_check:

File: srcCpp/perftest_publisher.cxx [hunk 37]
```diff
@@ -1843,15 +1449,17 @@ int perftest_cpp::Publisher()
...
-                                RTI_OSAPI_THREAD_PRIORITY_DEFAULT,
-                                RTI_OSAPI_THREAD_OPTION_DEFAULT,
-                                RTI_OSAPI_THREAD_STACK_SIZE_DEFAULT,
-                                NULL,
-                                LatencyReadThread,
-                                reader_listener);
+                    _PM.get<bool>("latencyTest") ? writer : NULL,
+                    _PM);
+
+            listenerThread = RTIOsapiThread_new(
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/perftest_publisher.cxx]: :ballot_box_with_check:

File: srcCpp/perftest_publisher.cxx [hunk 38]
```diff
@@ -2164,28 +1830,38 @@ int perftest_cpp::Publisher()
...
     }
 
-    if (_displayWriterStats) {
+    if (_PM.get<bool>("writerStats")) {
         printf("Pulled samples: %7d\n", writer->getPulledSampleCount());
     }
 
-    if (reader_listener != NULL) {
-        delete(reader_listener);
+    if (!finalize_thread(listenerThread, reader_listener)) {
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/perftest_publisher.cxx]: :ballot_box_with_check:

File: srcCpp/RTIDDSImpl.cxx [hunk 39]
```diff
@@ -1929,23 +1452,36 @@ class RTISubscriber : public IMessagingReader
...
+            if (status.current_count >= numPublishers) {
                 break;
             }
             perftest_cpp::MilliSleep(1000);
         }
     }
+
+    bool unblock()
+    {
+        DDSGuardCondition cond;
```
[the remark this answers is not in the document]
Comment (alfonso-soto) [reply srcCpp/RTIDDSImpl.cxx]: A condition will not be necessary. The waitset is already unblocking. (No complete sure about why. I am researching it.)
:ballot_box_with_check:

**Comment by alfonso-soto** [comment]
I check the performance degradation. All looks good. :ok_hand:

**Comment by javmorcas** [comment]
Close by #159

**Pull request closed without merging**